server/bleep/src/webserver/answer.rs

```diff
@@ -91,17 +93,70 @@ fn default_thread_id() -> uuid::Uuid {
     uuid::Uuid::new_v4()
 }
 
-pub(super) async fn handle(
-    Query(params): Query<Params>,
+pub(super) async fn answer(
+    Query(params): Query<Answer>,
     Extension(app): Extension<Application>,
     Extension(user): Extension<User>,
 ) -> super::Result<impl IntoResponse> {
     let query_id = uuid::Uuid::new_v4();
-    let response = _handle(
-        Query(params.clone()),
-        Extension(app.clone()),
-        Extension(user.clone()),
+
+    let conversation_id = ConversationId {
+        user_id: user
+            .login()
+            .ok_or_else(|| super::Error::user("didn't have user ID"))?
+            .to_string(),
+        thread_id: params.thread_id,
+    };
+
+    let (_, mut exchanges) = conversations::load(&app.sql, &conversation_id)
+        .await?
+        .unwrap_or_else(|| (params.repo_ref.clone(), Vec::new()));
+
+    let Answer {
+        parent_exchange_id,
+        q,
+        ..
+    } = &params;
+
+    if let Some(parent_exchange_id) = parent_exchange_id {
+        let truncate_from_index = if parent_exchange_id.is_nil() {
+            0
+        } else {
+            exchanges
+                .iter()
+                .position(|e| e.id == *parent_exchange_id)
+                .ok_or_else(|| super::Error::user("parent query id not found in exchanges"))?
+                + 1
+        };
+
+        exchanges.truncate(truncate_from_index);
+    }
+
+    let query = parser::parse_nl(q)
+        .context("parse error")?
+        .into_semantic()
+        .context("got a 'Grep' query")?
+        .into_owned();
+    let query_target = query
+        .target
+        .as_ref()
+        .context("query was empty")?
+        .as_plain()
+        .context("user query was not plain text")?
+        .clone()
+        .into_owned();
+
+    let action = Action::Query(query_target);
+    exchanges.push(Exchange::new(query_id, query));
+
+    let response = execute_agent(
+        params.clone(),
+        app.clone(),
+        user.clone(),
         query_id,
+        conversation_id,
+        exchanges,
+        action,
     )
     .await;
 
```

Comment:
By moving logic up and removing `_handle`, won't we now fail to track errors that are thrown by any logic outside of `execute_agent`? We want visibility over these errors.

Reply:
This is now patched in ca68093, by *always* tracking errors automatically in `execute_agent`.